[gskang] iP

gskang-22 opened the pull request.

gskang-22 committed:
Change chatbot name from Duke to XiaoMe

gskang-22 committed:
Implement an initial skeletal version

Let's create a chatbot that simnply greets the user and exits

gskang-22 committed:
Level1-1: implement echo functionality

XiaoMe now echos the input back to the user. XiaoMe exits when the user
types the command bye.

gskang-22 committed:
Level1-2: implement add and list commands

XiaoMe now stores whatever text entered by the user. The texts are
displayed back to the user when the user types the command list.

gskang-22 committed:
Extension A-Classes: Add new class Task to represent tasks

A task list is now represented using Task class instead of a
multi-dimensional array containing primitive values.

 This is a more natural approach and allows for further addition of
 features, such as task deadlines and priorities.

gskang-22 committed:
Level1-3: Implement the ability to mark tasks as done and undone

Users are now able to mark tasks as done/undone by typing the
command mark/unmark followed by an integer corresponding to the number
assigned to that task in the task list.

gskang-22 committed:
Update coding standard

Tweaked the code to comply with coding standard.

gskang-22 committed:
Create 3 new classes

Classes Deadline, Event, and Todo inherit from Task class. These classes
 are used to support multiple task types.

gskang-22 committed:
Add support for tracking 3 types of tasks: ToDos, Deadlines, Events

ToDos: tasks without any date/time attached to it
Deadlines: tasks that need to be done before a specific date/time
Events: tasks that start at a specific date/time and ends at a specific date/time

gskang-22 committed:
Add automated text UI testing

gskang-22 committed:
Improve code quality

Oshen27 commented on src/main/java/XiaoMe.java:
```diff
@@ -0,0 +1,154 @@
+        String line;
+        Scanner in = new Scanner(System.in);
+
+        System.out.println("""               
+                \t____________________________________________________________
+                \tHello! I'm XiaoMe
+                \tWhat can I do for you?
+                \t____________________________________________________________
+                """);
+
```
Should this be extracted out?

Oshen27 commented on src/main/java/XiaoMe.java:
```diff
@@ -0,0 +1,154 @@
+        BYE
+    }
+
+    public static Type checkType(String line) {
+        // checks what kind of command was received by XiaoMe
+
+        if (Objects.equals(line, "bye")) {
+            return Type.BYE;
+        } else if (Objects.equals(line, "list")) {
+            return Type.LIST;
+        }
+
+        String[] words = line.trim().split(" ");
+        String first = words[0];
+
+        if (Objects.equals(first, "todo")) {
+            return Type.TODO;
+        } else if (Objects.equals(first, "deadline")) {
+            return Type.DEADLINE;
+        } else if (Objects.equals(first, "event")) {
+            return Type.EVENT;
+        } else if (Objects.equals(first, "mark") || Objects.equals(first, "unmark")) {
+            return Type.MARK;
+        }
+
+        return Type.TASK;
+    }
```
Should this be in another file?

Oshen27 commented on src/main/java/Todo.java:
```diff
@@ -0,0 +1,10 @@
+    }
+
+    @Override
+    public String toString() {
+        return "[T]" + super.toString();
+    }
```
I like how you overrode the toString method

Oshen27 commented on src/main/java/XiaoMe.java:
```diff
@@ -0,0 +1,154 @@
+                        """);
+                break; // stop programme
+
+            } else if (type == Type.LIST) {
```
should you use switch case here?

Oshen27 commented on src/main/java/XiaoMe.java:
```diff
@@ -0,0 +1,154 @@
+                \t____________________________________________________________
+                """);
+
+        while (true) {
```
should this be extracted out?

yijiano commented on src/main/java/XiaoMe.java:
```diff
@@ -0,0 +1,154 @@
+            return Type.EVENT;
+        } else if (Objects.equals(first, "mark") || Objects.equals(first, "unmark")) {
+            return Type.MARK;
+        }
```
Should a `switch` statement with `case` clauses be used instead of `if-else` statements to improve readability?

yijiano commented on src/main/java/XiaoMe.java:
```diff
@@ -0,0 +1,154 @@
+        String line;
+        Scanner in = new Scanner(System.in);
+
+        System.out.println("""               
```
Should this be made into a function like `printLine()`?

yijiano commented on src/main/java/XiaoMe.java:
```diff
@@ -0,0 +1,154 @@
+                        \t____________________________________________________________
+                        \tBye. Hope to see you again soon!
+                        \t____________________________________________________________
+                        """);
```
Should this be extracted out?

yijiano commented on src/main/java/XiaoMe.java:
```diff
@@ -0,0 +1,154 @@
+        DEADLINE,
+        TASK,
+        BYE
+    }
```
I like the use of `enum` here to improve readability and usability.

yijiano commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    public String getStatusIcon() {
+        if (isDone) {
+            return "X";
+        } else {
+            return " ";
+        }
```
Should a ternary statement be used here instead?

yijiano reviewed (commented):
LGTM. I have a few nits with regards to code quality, but otherwise good job!

gskang-22 committed:
Handle error with try and catch
Prevents XiaoMe from crashing due to invalid user input by handling
these errors.

gskang-22 committed:
Add and use new class XiaoMeException
Use of exceptions to handle errors, and XiaoMeException to handle
invalid commands.

gskang-22 committed:
Move tasks-related classes to new package task
This improves the readability of the code.

gskang-22 committed:
Add support for deleting tasks from the list

Users are now able to delete tasks from the task list by using command "delete <index>".

gskang-22 committed:
Use of ArrayList for storing tasks data instead of Array

Increase ease of access of data due to the methods that come with ArrayList.

gskang-22 committed:
Implement save feature, where tasks are saved in txt file
Tasks are saved automatically in the hard disk  whenever the task list changes. When XiaoMe starts up, it loads the data from the hard disk.

gskang-22 committed:
Merge branch 'refs/heads/branch-Level-7'

# Conflicts:
#	src/main/java/XiaoMe.java

gskang-22 committed:
Fix bugs from merging

gskang-22 committed:
Create JAR file

Kobot7 commented on src/main/java/Save.java:
```diff
@@ -0,0 +1,73 @@
+import task.*;
```
Avoid using universal imports

Kobot7 commented on src/main/java/XiaoMe.java:
```diff
@@ -0,0 +1,259 @@
+                continue;
+            }
+            switch (type) {
+                case BYE:
```
Are the switch cases supposed to be indented?

Kobot7 commented on src/main/java/XiaoMe.java:
```diff
@@ -0,0 +1,259 @@
+            Type type;
+            try {
+                type = checkType(line);
+            } catch (XiaoMeException e) {
```
Is the variable name "e" informative enough?

Kobot7 commented on src/main/java/XiaoMe.java:
```diff
@@ -0,0 +1,259 @@
+                case LIST:
+
+                    // user input is list: display past tasks
+                    System.out.println("""
```
Consider extracting all your magic literals out

Kobot7 commented on src/main/java/XiaoMe.java:
```diff
@@ -0,0 +1,259 @@
+                            """);
+                continue;
+            }
+            switch (type) {
```
Consider extracting the functions for each switch case to another method to avoid deep nesting and long methods

Kobot7 replied:
Has all possible outcomes have been considered? May be best to include a default branch for all switch cases

Kobot7 commented on src/main/java/XiaoMe.java:
```diff
@@ -0,0 +1,259 @@
+                            throw new IllegalArgumentException();
+                        }
+
+                        tasks.add(new Event(eventWords[0].replace("event", "").trim(), eventWords[1].replace("from ", "").trim(), eventWords[2].replace("to ", "").trim())); // add task to storage
```
Consider splitting your code into multiple lines to ensure you do not exceed the line length limit

gskang-22 committed:
fix bugs with JAR file

gskang-22 committed:
fix bugs with delete command

gskang-22 committed:
Refactor code to extract out closely related code as classes

gskang-22 committed:
Merge remote-tracking branch 'refs/remotes/origin/master'

gskang-22 committed:
Refactor code to extract out closely related code as classes

gskang-22 committed:
Complete refactoring of code to extract out closely related code as classes (commands, parser, storage, task)

gskang-22 committed:
Implement Find feature
Give users a way to find a task by searching for a keyword.

gskang-22 committed:
Update error messages for find feature

gskang-22 committed:
Add JavaDoc comments

gskang-22 committed:
Merge pull request #1 from gskang-22/branch-Level-9

Branch level 9

gskang-22 committed:
Merge remote-tracking branch 'origin/branch-A-JavaDoc'

gskang-22 committed:
Add user guide in README.md

gskang-22 committed:
Improving user guide

gskang-22 committed:
Added TaskList.java class and reorganised files

gskang-22 committed:
Corrected some UI outputs

gskang-22 committed:
Update README.md

gskang-22 committed:
Update README.md

gskang-22 committed:
Update to fit ip requirements
